made the ball draggable with mouse : added grabbed property to Ball and _mouse_motion_cb to Bounce

marsian83 opened the pull request:
The game as of now allows motion via keyboard and accelometer. This PR aims to add support for grabbing the ball with the mouse and dragging it onto the answers. 

I have added `grabbed` property to the `Ball` class and have implemented a a callback for mouse motion
which checks if the ball is grabbed, move the ball to the mouse position

chimosky commented:
Reviewed 81fc056, not tested.

Your opening comment gives more context to the commit and should be part of your commit message.

quozl commented on bounce.py:
```diff
@@ -367,12 +368,15 @@ def _button_press_cb(self, win, event):
         ''' Callback to handle the button releases '''
         win.grab_focus()
         x, y = list(map(int, event.get_coords()))
+        self.ball.grabbed=False
```
`bounce.py:379:26: E225 missing whitespace around operator`

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
         return True
 
+    def _mouse_motion_cb(self, widget, event):
+        if self.ball.grabbed and self._fraction != None:
```
`bounce.py:398:49: E711 comparison to None should be 'if cond is not None:'`

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+                mouse_x, mouse_y, state = event.window.get_pointer()
+            else:
+                mouse_x = event.x
+                mouse_y = event.y
```
`bounce.py:403:17: F841 local variable 'mouse_y' is assigned to but never used`

You haven't thought this through?  :grin:

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+            else:
+                mouse_x = event.x
+                mouse_y = event.y
+                state = event.state
```
`bounce.py:404:17: F841 local variable 'state' is assigned to but never used`

Apparently not. :grinning:

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+                mouse_y = event.y
+                state = event.state
+
+                self.ball.move_ball_relative((mouse_x - self.ball.ball_x() - self.ball.width() / 2, 0))
```
`bounce.py:406:80: E501 line too long (103 > 79 characters)`

Oh dear.

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+                state = event.state
+
+                self.ball.move_ball_relative((mouse_x - self.ball.ball_x() - self.ball.width() / 2, 0))
+        
```
`bounce.py:407:1: W293 blank line contains whitespace`

Yay.

quozl commented on bounce.py:
```diff
@@ -367,12 +368,15 @@ def _button_press_cb(self, win, event):
         win.grab_focus()
         x, y = list(map(int, event.get_coords()))
         self._press = self._sprites.find_sprite((x, y))
+        if self._press is not None and self._press == self.ball.ball:
```
A bit strange looking, but what you are saying is if `self.ball.ball` is not None and `self._press` is the same.  Is there a better way to put this?

chimosky replied:
Agreed, you can just compare `self._press` and `self.ball.ball` as it'll fail if either is `None`.

quozl reviewed (changes requested):
Thanks, looking better.  Please also rewrite the commit messages as mentioned by @chimosky.

marsian83 commented:
@chimosky you ask me to change the commit messages so in order to do that should I force-push ? as this PR now contains flake8 fixes too

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+    def _mouse_motion_cb(self, widget, event):
+        if self.ball.grabbed and self._fraction is not None:
+            if event.is_hint:
+                mouse_x = event.window.get_pointer()[0]
```
On line 400, `mouse_x` is set but never used.  (Line 404 is indented within the false branch of the conditional test at line 399.)

quozl replied:
Calling `get_pointer` incurs a round-trip communication delay with the graphics system; is it really necessary to ascertain the actual position or is it sufficient to animate against the hinted position?

quozl commented on bounce.py:
```diff
@@ -390,6 +394,19 @@ def _button_release_cb(self, win, event):
+            else:
+                mouse_x = event.x
+
+                self.ball.move_ball_relative((mouse_x
```
Your new lines 405 and 406 trigger new flake8 warnings `W503 line break before binary operator`.

quozl reviewed (changes requested):
Thanks for the changes.

Please get into the habit of running flake8 and make sure that your new code does not introduce new warnings.  One method to do this is to run flake8 on the commit before your changes begin, save the output file, and run it again after your changes, then compare the output files.

Yes, we still want you to change the commit messages; our guide to contributing explains what we like to see, and we know you have said things in the pull request which we think should be in the commit message.

That your pull request now contains flake8 fixes is irrelevant, as every one of these fixes was to code that you introduced in the pull request.  If you doubt that, click on the _Files changed_ tab of the pull request to see the aggregation of the commits so far.

marsian83 commented on ball.py:
```diff
@@ -103,6 +103,8 @@ def __init__(self, sprites, filename):
         self.ball.set_layer(3)
         self.ball.set_label_attributes(24, vert_align='top')
 
+        self.grabbed = False
```
It just makes logical sense to me that the ball is the thing being grabbed so it could be it's property, if ball ever needs to access it in the future for, say a texture change when being grabbed, even then it could be used. If you think it would be better suited as an attribute of the activity class. please let me know, I will shift it to the activity class.

quozl replied:
Nothing in the class uses it.  That's the reason to change.  Having moved it to the activity class, it could be `drag`, `grab` or `mouse_grab` instead of `grabbed`.

marsian83 commented:
Hi @chimosky, I tried rebase and force push to rename my commit message but it seems to have pushed all the commits again, [14b0cc2](https://github.com/sugarlabs/fractionbounce/pull/17/commits/14b0cc2bcbc447e07e687e0a65d564acafaba87d) and all commits beyond that are duplicates. I do not know why this has happened. Do you have an idea? Can you help me out of this sticky situation?

quozl commented:
I see what you did.  It's a common mistake, nothing to worry about.  GitHub does not render the situation properly; this is a common criticism of GitHub's rendering.

What you have is a merge of two histories.  Here's what gitk shows me;

![gitk](https://user-images.githubusercontent.com/2936481/229982000-031f1cdb-3b3a-4a37-acce-154db38ecff4.png)

I'll next post how to resolve this quickly.

quozl commented:
A quick way to resolve this is to take your branch, reset the HEAD to the current master HEAD, then cherry-pick the commits on the most recent side of the merge.  Do this when you have half an hour available, in case you need to backtrack following an interruption.

The current master head is;
* e99ec8b ("v28.1")

The most recent commits you made were;
* 14b0cc2 ("Add support for grabbing the ball with the mouse")
* bd46b8d ("made the ball snap to the center of ball instead of sprite origin")
* 17e38b7 ("removed unnecessarily setting events")
* fc89a6a ("removed unneccesary comparison self.ball.ball is not None")
* 5ad94c1 ("resolved flake8 - missing whitespace around operator")
* c5e4783 ("resolved flake - comparison to None should be 'if cond is not None:'")
* 3ea6ae9 ("resolved flake8 - W293 blank line contains whitespace")
* 8563943 ("resolved flake8 - local variables assigned to but never used")
* 7ffbb6f ("resolved flake8 - line too long")
* 54caefa ("fixed inconsistent string quotes")
* a85b760 ("moved grabbed property of Ball class to mouse_grab of activity class")

I used the commit timestamps to make that assessment.  If my assessment of which commits you want is wrong, stop and tell me.  Otherwise proceed as follows.

First, make sure you have no uncommitted work on files, and if you do stash it.

Second, reset the branch;

```git
$ git reset --hard e99ec8b
```

Third, cherry-pick the commits you want to keep;

```git
$ git cherry-pick \
14b0cc2bcbc447e07e687e0a65d564acafaba87d \
bd46b8da410f4b1f6b4fe06cb87f5452cd068e5f \
17e38b7be3c26533de7c32fa840659940846190f \
fc89a6ae2549d756fa1ce40e617a89c225de40ba \
5ad94c184c9204c1ed58114d5f9b8390df0d5bff \
c5e47835f37755f2f9cb2889cd86822539b0e476 \
3ea6ae9f5657dc583d6d0a21d25db6c5e27f4c94 \
8563943b786ab8ad20aa9a2614bb020c63d321ec \
7ffbb6f4a1bc7a8fce3e4a88da40c1e19221436f \
54caefa220e287bbdc93f0fcb7748a8736490238 \
a85b760982e7dbf958492bea727b423c24aeab24
```

Fourth, push the branch and the pull request will be updated;
```git
$ git push --force
```

Should you need to undo this, you can reset your branch to the merge commit and push again;

```git
$ git reset --hard 0d67b0d
$ git push --force
```

quozl commented:
Yay.

quozl commented:
If it were me, I'd use `git rebase -i` to squash the flake8 commits into the commits that introduced the flake8 excursions.  Remember you can always _reset_ your branch to what it was if you make a note of the commit.

marsian83 commented:
Hi, thanks a lot for the help with the commits. Also, I have changed the `_mouse_motion_cb` function and also moved the `grabbed` attribute of Ball class to Bounce and renamed it to `mouse_grab`

marsian83 commented on bounce.py:
```diff
@@ -367,12 +368,15 @@ def _button_press_cb(self, win, event):
         x, y = list(map(int, event.get_coords()))
         self._press = self._sprites.find_sprite((x, y))
+        if self._press == self.ball.ball:
+            self.ball.grabbed = True
```
considering my options 🤔

chimosky commented:
Reviewed 3937d14 and your commit message says "- After testing, working without accurate position is enough", could you include why it's enough as you say your tests tell you it's enough but that's vague as it's your tests and not anyone else's.

marsian83 commented:
By testing it, I mean I tried it and noticed not too big of a difference in the movement of the ball relative to the mouse even when working with the hinted position compared to the accurate position. I'm not sure as to what explanation to include in the commit message?

chimosky commented:
> By testing it, I mean I tried it and noticed not too big of a difference in the movement of the ball relative to the mouse even when working with the hinted position compared to the accurate position. I'm not sure as to what explanation to include in the commit message?

I understand you now, no need to change the commit message.
